spec-tests/src/test/java/org/candlepin/spec/pools/RefreshPoolsTest.java

```diff
@@ -0,0 +1,877 @@
+            owner.getKey(), null, true);
+            ApiClient consumerClient = ApiClients.trustedConsumer(user.getUuid());
+
+            JsonNode entNode = consumerClient.consumers().bindPool(user.getUuid(), pool.getId(), 1);
+            assertEquals(1, entNode.size());
+
+            user = adminClient.consumers().getConsumer(user.getUuid());
+            assertEquals(1, user.getEntitlementCount());
+        }
+
+        // Verify the entitlement count for this pool
+        List<EntitlementDTO> ents = adminClient.pools().getPoolEntitlements(pool.getId());
+        assertEquals(sub.getQuantity(), ents.size());
+
+        // Modify the subscription upstream
+        sub.setQuantity(1L);
+        adminClient.hosted().updateSubscription(sub.getId(), sub);
+
+        refreshPools(adminClient, owner.getKey());
+        pools = adminClient.pools().listPoolsByOwner(owner.getId());
+        assertEquals(1, pools.size());
+        pool = pools.get(0);
+        assertEquals(sub.getQuantity(), pool.getQuantity());
+
+        // Verify the entitlement count has changed
+        ents = adminClient.pools().getPoolEntitlements(pool.getId());
+        assertEquals(1, ents.size());
+    }
+
+    private AsyncJobStatusDTO refreshPools(ApiClient client, String ownerKey) {
+        AsyncJobStatusDTO job = adminClient.owners().refreshPools(ownerKey, true);
+        assertNotNull(job);
+        job = adminClient.jobs().waitForJob(job);
+        assertEquals("FINISHED", job.getState());
+
+        return job;
+    }
+
+    private void compareProducts(ProductDTO expected, ProductDTO actual) {
+        assertEquals(expected.getId(), actual.getId());
+        assertEquals(expected.getCreated(), actual.getCreated());
+        assertEquals(expected.getUpdated(), actual.getUpdated());
+        assertEquals(expected.getName(), actual.getName());
+        assertEquals(expected.getMultiplier(), actual.getMultiplier());
+        assertEquals(expected.getAttributes(), actual.getAttributes());
+    }
+
+    private void compareBrandings(BrandingDTO expected, BrandingDTO actual) {
+        assertEquals(expected.getProductId(), actual.getProductId());
+        assertEquals(expected.getName(), actual.getName());
+        assertEquals(expected.getType(), actual.getType());
+    }
+
+    private void compareBrandings(Collection<BrandingDTO> expected, Collection<BrandingDTO> actual) {
+        assertEquals(expected.size(), actual.size());
+        for (BrandingDTO expectedBrand : expected) {
+            boolean found = false;
+            for (BrandingDTO actualBrand : actual) {
+                if (expectedBrand.getProductId().equals(actualBrand.getProductId())) {
+                    compareBrandings(expectedBrand, actualBrand);
+                    found = true;
+                }
+            }
+
+            assertTrue(found, "missing branding " + expectedBrand.getName());
+        }
+    }
+
+    private void verifyCertProductsAndContent(JsonNode certBody,
```

Comment:
I went back and forth on this when writing the method, and I like the idea of keeping it a map in case we end up adding tests that want to check content for multiple products.

Reply:
I think in practice, setting up the map will be more work than calling the method multiple times. In terms of setting this up to be used in other places/tests going forward, we'd get more value flipping the logic here and converting the JsonNode into a map of pids-to-cids, and then processing it in-test:

```
private static Map<String, List<String>> toProductContentIdMap(JsonNode certNode) {
  // TODO: add reasonable node assertions or checks
  Map<String, List<String>> output = new HashMap<>();

  JsonNode products = certNode.get("products");
  
  products.forEach(productNode -> {
    String productId = productNode.get("id").asText();

    // note that the json could very well list the same content multiple times, 
    // so we can't safely use a set here.
    List<String> contentIds = output.computeIfAbsent(productId, () -> new ArrayList<>()); 

    content = productNode.get("content");
    content.forEach(contentNode -> contentIds.add(contentNode.get("id").asText()));
  })

  return output;
}

...

Map<String, List<String>> certProductContentIds = toProductContentIdMap(certNode);
assertThat(certProductContentIds).containsOnly(
  entry("expected pid1", List.of("expected", "content", "ids")),
  entry("expected pid2", List.of("expected cid")),
  entry("expected pid3", List.of("expected cid 2")));

// Or, for the more verbose among us:
assertThat(certProductContentIds)
  .isNotNull()
  .hasSize(3)
  .containsOnlyKeys("expected", "product", "ids");

assertThat(certProductContentIds.get("expected"))
  .isNotNull()
  .hasSize(3)
  .containsOnly("expected", "content", "ids");
```

Not a big deal either way, just something to consider going forward. The more complex and context-specific the input is to a utility method, the less likely it is to be used as intended, or at all.

Reply:
I like having a method that returns the map of product id to content ids and asserting against the map in the test method a lot more than the method I had! Updated